staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go

```diff
@@ -255,6 +255,113 @@ func TestTimeoutHeaders(t *testing.T) {
 	res.Body.Close()
 }
 
+func TestTimeoutRequestHeaders(t *testing.T) {
+	origReallyCrash := runtime.ReallyCrash
+	runtime.ReallyCrash = false
+	defer func() {
+		runtime.ReallyCrash = origReallyCrash
+	}()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Add dummy request info, otherwise we skip postTimeoutFn
+	ctx = request.WithRequestInfo(ctx, &request.RequestInfo{})
+
+	withDeadline := func(handler http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			handler.ServeHTTP(w, req.WithContext(ctx))
+		})
+	}
+
+	ts := httptest.NewServer(
+		withDeadline(
+			WithTimeoutForNonLongRunningRequests(
+				http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+					// trigger the timeout
+					cancel()
+					// mutate request Headers
+					// Authorization filter does it for example
+					for j := 0; j < 10000; j++ {
+						req.Header.Set("Test", "post")
+					}
+				}),
+				func(r *http.Request, requestInfo *request.RequestInfo) bool {
+					return false
+				},
+			),
+		),
+	)
+	defer ts.Close()
+
+	client := &http.Client{}
+	req, err := http.NewRequest(http.MethodPatch, ts.URL, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	res, err := client.Do(req)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if res.StatusCode != http.StatusGatewayTimeout {
+		t.Errorf("got res.StatusCde %d; expected %d", res.StatusCode, http.StatusServiceUnavailable)
+	}
+	res.Body.Close()
+}
+
+func TestTimeoutWithLogging(t *testing.T) {
```

Comment:
I can't trigger the race with this test :thinking: 
```
 stress ./filters.test -test.run TestTimeoutWithLogging
5s: 744 runs so far, 0 failures
10s: 1501 runs so far, 0 failures
15s: 2263 runs so far, 0 failures
20s: 3022 runs so far, 0 failures
25s: 3780 runs so far, 0 failures
30s: 4538 runs so far, 0 failures
35s: 5291 runs so far, 0 failures
40s: 6044 runs so far, 0 failures
45s: 6788 runs so far, 0 failures
50s: 7552 runs so far, 0 failures
55s: 8303 runs so far, 0 failures
1m0s: 9055 runs so far, 0 failures
1m5s: 9810 runs so far, 0 failures
```

without the fix , of course

Reply:
ok, in `TestTimeoutRequestHeaders` it fails inmidiatly

Reply:
In case of TestTimeoutWithLogging passing --args -v=3 is necessary to force logging. There is a condition in the code, that we only run deferred logging, when klog verbosity is >= 3